src/generate/context.rs

```diff
@@ -228,25 +200,92 @@ impl CppContext {
         }
 
         println!("Writing {:?}", self.typedef_path.as_path());
-        let mut writer = CppWriter {
+        let mut typedef_writer = CppWriter {
             stream: File::create(self.typedef_path.as_path())?,
             indent: 0,
         };
+        let _typeimpl_writer = CppWriter {
+            stream: File::create(self.type_impl_path.as_path())?,
+            indent: 0,
+        };
+        let _fundamental_writer = CppWriter {
+            stream: File::create(self.fundamental_path.as_path())?,
+            indent: 0,
+        };
+
         // Write includes for typedef
         self.typedef_includes
             .iter()
-            .for_each(|inc| inc.write(&mut writer).unwrap());
-        self.declarations
-            .iter()
-            .for_each(|dec| dec.write(&mut writer).unwrap());
-        self.types.iter().for_each(|(k, v)| {
-            writeln!(writer, "/* {:?} */", k).unwrap();
-            v.write(&mut writer).unwrap();
+            .for_each(|inc| inc.write(&mut typedef_writer).unwrap());
+
+        // forward declares
+        for (namespace, strings) in &self.typedef_declarations {
+            writeln!(typedef_writer, "namespace {} {{", namespace)?;
+            strings
+                .iter()
+                .for_each(|s| s.write(&mut typedef_writer).unwrap());
+            writeln!(typedef_writer, "}} // namespace {}", namespace)?;
+        }
+        self.typedef_types.iter().for_each(|(k, v)| {
+            writeln!(typedef_writer, "/* {:?} */", k).unwrap();
+            v.write(&mut typedef_writer).unwrap();
         });
 
         // TODO: Write type impl and fundamental files here
         Ok(())
     }
+
+    fn fill_type(
+        &mut self,
+        metadata: &Metadata,
+        config: &GenerationConfig,
+        ctx_collection: &mut CppContextCollection,
+        ty: TypeData,
+    ) {
+        let tag = TypeTag::from(ty);
+        if let TypeData::TypeDefinitionIndex(tdi) = ty {
+            let cpp_type_entry = self.typedef_types.entry(tag);
+
+            if let Entry::Occupied(occupied) = cpp_type_entry {
+                let mut cpp_type = occupied.remove();

```

Comment:
Why do we have to remove it just to add it back? Seems potentially problematic.

Reply:
Rust borrow checker. It removes to take ownership because rust doesn't like mutable borrow (and I need to pass &mut self as a param)